lib/opal/nodes/literal.rb

```diff
@@ -213,42 +213,132 @@ class XStringNode < Base
-            # we have to use raw source
-            # so we need to combine "return" with "raw_source"
-            push 'return '
-            str = child.children.first
-            value = str.loc.expression.source
-            push Fragment.new(value, nil)
+
+            should_add_semicolon = true if returning
+          end
+
+          if single_child
+            push Fragment.new(last_value, scope, last_child)
           else
-            raise "Unsupported xstr part: #{child.type}"
+            compile_child(first_child)
+            stripped_children.each { |c| compile_child(c) }
+            if last_child.type == :str
+              should_add_semicolon = false if first_child.type != :str
+              push Fragment.new(last_value, scope, last_child)
+            else
+              compile_child(last_child)
+            end
           end
+
+        else
+          # Here we leave to the user the responsibility to add
+          # a return where it's due.
+          children.each { |c| compile_child(c) }
         end
 
         wrap '(', ')' if recv?
+        push ';' if should_add_semicolon
       end
 
-      def detect_and_remove_trailing_semicolon(value, node)
-        if value =~ /;\s*#{REGEXP_END}/
-          compiler.warning 'Do not terminate one-line xstr expression with semicolon', node.line
-          value.sub(/;\s*#{REGEXP_END}/, '')
+      def compile_child(child)
+        case child.type
+        when :str
+          value = child.loc.expression.source
+          push Fragment.new(value, scope, child)
+        when :begin, :gvar, :ivar
+          push expr(child)
         else
-          value
+          raise "Unsupported xstr part: #{child.type}"
+        end
+      end
+
+      # Check if there's only one child or if they're all part of
+      # the same line (e.g. because of interpolations)
+      def single_line?(children)
+        (children.size == 1) ||
+          children.none? do |c|
+            c.type == :str && c.loc.expression.source.end_with?("\n")
+          end
+      end
+
+      # A case for manually created :js_return statement in Compiler#returns
+      # Since we need to take original source of :str we have to use raw source
+      # so we need to combine "return" with "raw_source"
+      def unpack_return(children)
+        first_child = children.first
+        returning   = false
+
+        if first_child.type == :js_return
+          returning = true
+          children = first_child.children
         end
+
+        return returning, children
+      end
+
+      # Will remove empty :str lines coming from cosmetic newlines in x-strings
+      #
+      # @example
+      #   # this will generate two additional empty
+      #   # children before and after `foo()`
+      #   %x{
+      #     foo()
+      #   }
+      def strip_empty_children(children)
+        start_index = 0
+        end_index = children.size - 1
+
+        while start_index <= end_index &&
```

Comment:
At least the first loop definitely could be replaced to

``` ruby
children.drop_while { |child| <while condition> }
```

But there's no `pop_while` or any similar method. Maybe something like:
``` ruby
_, *children_to_keep, _ = children.chunk { |child| <while condition> }
```

Reply:
That's true for Opal, but what about MRI? `opal-parser` is slow anyway 😄 . And from what I understand `opal-parser` is just a proof-of-concept, everyone uses Ruby-based version of Opal.

Reply:
Flip-flop solves it very well by they way :)